README.MD
# HTTP server port

This port exposes the following functionality:

1. Plain HTTP/HTTPS for static resources
1. GET/POST requests with JSON-RPC 2.0 body to the **/rpc** URL
1. GET/POST requests with JSON body equivalent of the JSON-RPC 2.0 "params" property to the **/rpc/namespace/method** URL
1. There are some predefined namespaces and their methods:
  * identity (equivalent to security)
    * **identity.login** - create a session
    * **identity.logout** - close a session
    * **identity.changePassword** - change password
  * permission
    * **permission.check** - check if a permission for an action is granted
    * **permission.list** - list permitted actions

# Planned features

1. Secure Remote Password protocol support, as defined in [RFC 2945](http://tools.ietf.org/html/rfc2945)
1. Multi language support
1. Fingerprinted resource URLs
1. WebSockets
1. Recommendations from http://stackoverflow.com/questions/549/the-definitive-guide-to-form-based-website-authentication

# Planned node modules to be researched

1. For debugging: https://www.npmjs.com/package/tv
1. For caching: https://www.npmjs.com/package/catbox
1. For WebSockets: https://www.npmjs.com/package/shoe

## USAGE

#### Configuration file required options:

```javascript
    id: '',
    type: 'httpserver',
    logLevel: '',
    port: ''
```

* `id`: unique identification of port

* `logLevel`: trace, debug or info.

* `port`: the port where the server will accept connections.

* `ports`: same as port but sets multiple listening ports

* `server`: Server specific configuration

* `fileUpload`: File upload configuration

server specific configuration can be passed as object with key `server`.
For instance if we have http server port implementation as follows:
```javascript
module.exports = {
    id: 'httpserver',
    type: 'httpserver',
    logLevel: 'trace',
    port: 8003,//or use property ports, see below
    ports: [8003, 8004],
    imports:[],
    jwt: { // jwt options (https://github.com/auth0/node-jsonwebtoken)
        verifyOptions: {...}, // options that are used when verifying
        signOptions: {...}, // options that are used when signing
        key: '...' // jwt sign key
    },
    disableXsrf: {http: false, ws: false}, // disable xsrf support for http and ws(web sockets)
    disablePermissionVerify: {ws: false}, // disable verification of services, eg pass requests without checks
    setSecurityHeaders: false, // enable security headers for every request. Usefull when this headers are not set in reverse proxy
    ssoAuthUrl: '...', // to where client should be redirected in order to make single sign on authentication
    appId: 'appName', // required for permission verify in socket-server
    identityNamespace: 'identity', // default value: identity, if identityNamespace is set to false, no identity check will be applied. this is useful when other namespace than identity will be used for identity check
    cookiePaths:'/rpc', // cookie paths, to which paths to bound cookies to, default: /rpc
    fileUpload: {
        maxFileName: 100,
        payloadMaxBytes: 5242880, // 5 MB. Default is 1048576 (1MB)
        extensionsWhiteList: ['pdf', 'doc', 'docx', 'xls', 'txt', 'jpg', 'jpeg', 'png']
    },
    validationPassThrough: (true|false), // if true, validation is not mandatory for methods. default policy: restrictive
    entryPoint: './abc/index.js', // app entry point (webpack helper)
    appId: 'monitoring', // will help when socket server matches permissions.
    server:{
        /*.......*/
    },
    registry: {
        name: 'httpserver007',
        host: '17.0.0.1',
        port: '8030',
        context: {
            key1: 'value1',
            key2: 'value2'
        }
    }
    /*.......*/
};
```
all properties from `server` will be passed to server as configuration and will be merged with `host` and `port` properties

* `registry` this is an optional property instructing how the http server should regiter itself as a service in case service discovery have been enabled for the implementation (for more information about how to enable service discovery on implementation level can be found [here](https://github.com/softwaregroup-bg/ut-port-registry).
This parameter is entirely optional. If not provided then the following default values will be used:
```
    {
        name: 'the name of the implementation',
        host: 'the host the http server is started on',
        port: 'the port the http server is started on',
        context: {
            type: 'http'
        }
    }
```
If only the name needs to be changed from the default input then the following configuration must be passed:

```
    registry: {
        name: 'custom name'
    }
```
To disable service registration set
```
{
    registry: false
}
```

Optional configuration options:
* `start`: function that will be called once on creation of port. It can be used for setting global variables and initializing objects.

## Validations (config) files pops

```javascript
module.exports = {
    description: '...', // from swagger description
    notes: ['...'], // from swagger notes
    tags: ['...'], // from swagger tags, it will add api automatically
    isRpc: false || true, // either this route is rpc or not
    auth: false || '....', // authentication type (jwt ?)
    httpMethod: '..', // (GET|POST|PUT|DELETE)
    response: '...',// response and result, bot are used for response validation, if both are omitted response will not bi validated!
    result: '...',
    route: '<route path>' // example "/a/b/c/{page}" will overwrite default defined route path
};
```

## Return streamed file (used for large files)
if `staticFileName` is set file will be returned to client, but if `tmpStaticFileName` is set after file gets returned will be deleted.
Downloaded file name could be set explicitly with `downloadFileName`, otherwise `staticFileName` or `tmpStaticFileName` would be used as default.


CHANGELOG.md
## [11.0.1](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v11.0.0...v11.0.1) (2020-07-09)



# [11.0.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.8.5...v11.0.0) (2020-04-30)


### Bug Fixes

* invalid path ([8cd17c7](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/8cd17c7e0a45ad4249e2bdca779c3ed714e6a30d))


### Features

* return specific error when validation fails ([1406c99](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/1406c9991b5a1c395a31231bb7cb1204f64bc075))



## [10.8.5](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.8.4...v10.8.5) (2020-01-28)


### Bug Fixes

* do not break api docs due to missing services ([10fed56](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/10fed56))



## [10.8.4](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.8.3...v10.8.4) (2020-01-24)



## [10.8.3](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.8.2...v10.8.3) (2020-01-17)



## [10.8.2](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.8.1...v10.8.2) (2020-01-16)


### Bug Fixes

* call imported methods using port namespace ([#151](https://github.com/softwaregroup-bg/ut-port-httpserver/issues/151)) ([0353901](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/0353901))



## [10.8.1](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.8.0...v10.8.1) (2019-12-02)


### Bug Fixes

* update ut-api ([a851ebb](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/a851ebb))



# [10.8.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.8.0-api.5...v10.8.0) (2019-12-02)


### Bug Fixes

* update ut-api ([e01d979](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/e01d979))



## [10.7.1](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.8.0-api.0...v10.7.1) (2019-11-25)


### Bug Fixes

* reuse all validations when overriding the identity http routes ([#150](https://github.com/softwaregroup-bg/ut-port-httpserver/issues/150)) ([60707f5](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/60707f5))



## [10.7.1](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.7.0...v10.7.1) (2019-11-25)


### Bug Fixes

* reuse all validations when overriding the identity http routes ([#150](https://github.com/softwaregroup-bg/ut-port-httpserver/issues/150)) ([60707f5](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/60707f5))



# [10.7.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.6.0...v10.7.0) (2019-08-02)


### Features

* support address in config ([2973aca](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/2973aca))



# [10.6.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.5.2...v10.6.0) (2019-07-19)


### Features

* expose k8s config ([7f7e35b](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/7f7e35b))



## [10.5.2](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.5.1...v10.5.2) (2019-04-16)


### Bug Fixes

* remove unnecessary logging ([a59d151](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/a59d151))



## [10.5.1](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.5.0...v10.5.1) (2019-04-15)


### Bug Fixes

* allow channel in auth ([c1f8373](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/c1f8373))



# [10.5.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.4.6...v10.5.0) (2019-04-10)


### Features

* added channel to the JWT ([#144](https://github.com/softwaregroup-bg/ut-port-httpserver/issues/144)) ([1d44d02](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/1d44d02))



## [10.4.6](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.4.5...v10.4.6) (2019-03-22)



## [10.4.5](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.4.4...v10.4.5) (2019-03-14)


### Bug Fixes

* set timer only when there is method ([0b8e54e](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/0b8e54e))



## [10.4.4](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.4.3...v10.4.4) (2019-03-07)



## [10.4.3](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.4.2...v10.4.3) (2019-03-01)


### Bug Fixes

* set abortEarly to false ([27e7150](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/27e7150))



## [10.4.2](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.4.1...v10.4.2) (2019-02-23)


### Bug Fixes

* update jsonwebtoken ([8655f99](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/8655f99))



## [10.4.1](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.4.0...v10.4.1) (2019-02-19)


### Bug Fixes

* preserve import order ([ffec442](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/ffec442))



# [10.4.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.3.0-rc-godfather.11...v10.4.0) (2019-02-02)



# [10.3.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.3.0-rc-godfather.1...v10.3.0) (2018-09-21)



<a name="10.3.0"></a>
# [10.3.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.3.0-rc-godfather.0...v10.3.0) (2018-09-21)


### Bug Fixes

* global health check + cleanup ([487ea64](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/487ea64))



<a name="10.2.0"></a>
# [10.2.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.2.0-rc-einstein.5...v10.2.0) (2018-08-02)



<a name="10.1.0"></a>
# [10.1.0](https://github.com/softwaregroup-bg/ut-port-httpserver/compare/v10.0.0...v10.1.0) (2018-06-07)


### Bug Fixes

* timeout validation ([#122](https://github.com/softwaregroup-bg/ut-port-httpserver/issues/122)) ([db556c4](https://github.com/softwaregroup-bg/ut-port-httpserver/commit/db556c4))



<a name="10.0.0"></a>

index.js
'use strict';
const path = require('path');
const util = require('util');
const hapi = require('hapi');
const inert = require('inert');
const vision = require('vision');
const jwt = require('hapi-auth-jwt2');
const basicAuth = require('hapi-auth-basic');
const mergeWith = require('lodash.mergewith');
const swagger = require('hapi-swagger');
const packageJson = require('./package.json');
const handlers = require('./handlers');
const fs = require('fs-plus');
const SocketServer = require('./socketServer');
const uuid = require('uuid/v4'); 
const hoek = require('hoek');
const Boom = require('boom');
require('./fix-swagger');
let errors;
module.exports = function({parent}) {
    function HttpServerPort({config}) {
        parent && parent.apply(this, arguments);
        this.config = mergeWith({
            id: null,
            logLevel: 'info',
            type: 'httpserver',
            port: 8080,
            connections: [],
            identityNamespace: 'identity',
            routes: {
                rpc: {
                    method: 'POST',
                    path: '/rpc/{method?}',
                    config: {
                        auth: 'jwt',
                        payload: {
                            output: 'data',
                            parse: true
                        }
                    }
                }
            },
            publicMethods: [
                'identity.forgottenPasswordRequest',
                'identity.forgottenPasswordValidate',
                'identity.forgottenPassword',
                'identity.registerRequest',
                'identity.registerValidate'
            ],
            cookie: {
                ttl: 100 * 60 * 1000, // expires a year from today
                encoding: 'none', // we already used JWT to encode
                isSecure: true, // warm & fuzzy feelings
                isHttpOnly: true, // prevent client alteration
                clearInvalid: false, // remove invalid cookies
                strictHeader: true // don't allow violations of RFC 6265
            },
            cookiePaths: '/rpc',
            disableXsrf: {http: false, ws: false}, // disable xsrf support for http and ws(web sockets)
            disablePermissionVerify: {ws: false}, // disable verification of services, eg pass requests without checks
            setSecurityHeaders: false,
            fileUpload: {
                maxFileName: 100,
                payloadMaxBytes: 5242880, // 5 MB. Default is 1048576 (1MB)
                extensionsWhiteList: ['pdf', 'doc', 'docx', 'xls', 'txt', 'jpg', 'jpeg', 'png']
            },
            swagger: {
                info: {
                    version: packageJson.version
                },
                pathPrefixSize: 2 // this helps extracting the namespace from the second argument of the url
            },
            jwt: {
                cookieKey: 'ut5-cookie',
                key: 'ut5-secret',
                verifyOptions: {
                    ignoreExpiration: true,
                    algorithms: ['HS256']
                },
                signOptions: {
                    expiresIn: '1h',
                    algorithm: 'HS256'
                }
            }
        }, config);
        errors = errors || require('./errors')(this.defineError);
        this.hapiServer = {};
        this.socketServer = null;
        this.socketSubscriptions = [];
        this.routes = [];
        this.stream = {};
    }

    if (parent) {
        util.inherits(HttpServerPort, parent);
    }

    HttpServerPort.prototype.init = function init() {
        parent && parent.prototype.init.apply(this, arguments);
        this.latency = this.counter && this.counter('average', 'lt', 'Latency');
        this.bytesSent = this.counter && this.counter('counter', 'bs', 'Bytes sent', 300);
        this.bytesReceived = this.counter && this.counter('counter', 'br', 'Bytes received', 300);
        this.hapiServer = new hapi.Server();
        this.bus.registerLocal({
            registerRequestHandler: this.registerRequestHandler.bind(this)
        }, this.config.id);
    };

    HttpServerPort.prototype.start = function start() {
        this.bus && this.bus.importMethods(this.config, this.config.imports, undefined, this);
        let args = Array.prototype.slice.call(arguments);
        this.context = {requests: {}};
        this.stream = this.pull(false, this.context);

        let captureMetrics = connection => {
            connection.listener.on('connection', socket => {
                socket.on('data', packet => {
                    packet && packet.length && this.bytesReceived && this.bytesReceived(packet.length);
                });
                let write = socket.write;
                socket.write = (data, encoding, callback) => {
                    return write.call(socket, data, encoding, (...params) => {
                        this.bytesSent && this.bytesSent(Buffer.byteLength(data, encoding));
                        callback && callback(...params);
                    });
                };
            });
        };

        if (this.config.connections && this.config.connections.length) {
            this.config.connections.forEach((connection) => {
                captureMetrics(this.hapiServer.connection(Object.assign({port: (this.config.port == null) ? 8080 : this.config.port}, connection)));
            });
        } else {
            captureMetrics(this.hapiServer.connection({
                port: (this.config.port == null) ? 8080 : this.config.port
            }));
        }

        if (this.config.setSecurityHeaders) {
            this.hapiServer.ext('onPreResponse', function(request, reply) {
                if (request.response && request.response.header) {
                    request.response.header('X-Content-Type-Options', 'nosniff');
                    request.response.header('X-Frame-Options', 'SAMEORIGIN');
                    request.response.header('Strict-Transport-Security', 'max-age=31536000; includeSubDomains');
                }
                return reply.continue();
            });
        }

        return new Promise((resolve, reject) => {
            let fileUploadTempDir = path.join(this.bus.config.workDir, 'ut-port-httpserver', 'uploads');
            fs.access(fileUploadTempDir, fs.R_OK | fs.W_OK, function(err) {
                if (err) {
                    if (err.code === 'ENOENT') {
                        fs.makeTree(fileUploadTempDir, function(err) {
                            if (err) {
                                reject(err);
                            } else {
                                resolve('Temp dir for file uploads has been created: ' + fileUploadTempDir);
                            }
                        });
                    } else {
                        reject(err);
                    }
                } else {
                    resolve('Temp dir for file uploads has been verified: ' + fileUploadTempDir);
                }
            });
        }).then((dir) => {
            return new Promise((resolve, reject) => {
                this.hapiServer.register([
                    basicAuth,
                    jwt,
                    inert,
                    vision, {
                        register: swagger,
                        options: this.config.swagger
                    }
                ], (e) => (e ? reject(e) : resolve()));
            });
        })
        .then(() => {
            this.hapiServer.auth.strategy('basic', 'basic', {
                validateFunc: (request, username, password, cb) => {
                    cb(null, true, {username: username, password: password});
                }
            });
            this.hapiServer.auth.strategy('jwt', 'jwt', true, mergeWith({
                validateFunc: (decoded, request, cb) => (cb(null, true)) // errors will be matched in the rpc handler
            }, this.config.jwt));
            return 0;
        })
        .then(() => {
            this.hapiServer.route(this.routes);
            this.hapiServer.route(handlers(this, errors));
            return 0;
        })
        .then(() => parent && parent.prototype.start.apply(this, args))
        .then(() => {
            if (this.socketSubscriptions.length) {
                this.socketServer = new SocketServer(this, this.config);
                this.socketSubscriptions.forEach((config) => this.socketServer.registerPath.apply(this.socketServer, config));
                this.socketServer.start(this.hapiServer.listener);
            }
            return 0;
        })
        .then(() => new Promise((resolve, reject) => {
            this.hapiServer.start((e) => {
                if (e) {
                    return reject(e);
                } else if (this.bus.config.registry && this.config.registry !== false) {
                    this.hapiServer.route({
                        method: 'GET',
                        path: '/health',
                        config: {
                            auth: false,
                            handler: (request, reply) => {
                                return this.isReady ? reply('ok') : reply('service not available').code(503);
                            }
                        }
                    });
                    let info = this.hapiServer.info;
                    let config = mergeWith(
                        // defaults
                        {
                            name: this.bus.config.implementation,
                            address: info.host, // info.address is 0.0.0.0 so we use the host
                            port: info.port,
                            id: uuid(),
                            check: {},
                            context: {
                                type: 'http',
                                pid: process.pid
                            }
                        },
                        // custom
                        this.config.registry
                    );
                    config.check.http = `${config.protocol || info.protocol}://${config.address}:${config.port}/health`;
                    return this.bus.importMethod('registry.service.add')(config)
                        .then(resolve)
                        .catch(reject);
                }
                return resolve();
            });
        }))
        .then(() => {
            this.log.info && this.log.info({
                message: 'HTTP server started',
                $meta: {
                    mtid: 'event',
                    opcode: 'port.started'
                }
            });
            return true;
        });
    };

    HttpServerPort.prototype.registerRequestHandler = function(handlers) {
        if (this.hapiServer.route && this.hapiServer.connections.length) {
            this.hapiServer.route(handlers);
        } else {
            Array.prototype.push.apply(this.routes, (handlers instanceof Array) ? handlers : [handlers]);
        }
    };

    HttpServerPort.prototype.registerSocketSubscription = function(path, verifyClient, opts) {
        this.socketSubscriptions.push([path, verifyClient, opts]);
        return (params, message) => this.socketServer.publish({path: path, params: params}, message);
    };

    HttpServerPort.prototype.enableHotReload = function enableHotReload(config) {
        return new Promise((resolve, reject) => {
            if (this.hotReload) {
                resolve(true);
            } else if (this.config.packer && this.config.packer.name === 'webpack') {
                let webpack = require('webpack');
                if (typeof config.output !== 'object') {
                    return reject(new Error('config.output must be an Object'));
                }
                if (typeof config.output.publicPath !== 'string') {
                    return reject(new Error('config.output.publicPath must be a String'));
                }
                if (typeof config.entry !== 'object') {
                    return reject(new Error('config.entry must be an Object'));
                }
                for (let name in config.entry) {
                    if (config.entry.hasOwnProperty(name) && name !== 'vendor') {
                        if (!Array.isArray(config.entry[name])) {
                            return reject(new Error(config.entry[name] + ' should be an Array'));
                        }
                        (config.entry[name].indexOf('webpack-hot-middleware/client') < 0) && config.entry[name].unshift('webpack-hot-middleware/client');
                        (config.entry[name].indexOf('react-hot-loader/patch') < 0) && config.entry[name].unshift('react-hot-loader/patch');
                    }
                }
                if (!Array.isArray(config.plugins)) {
                    return reject(new Error('config.plugins must be an Array'));
                }
                config.plugins.push(new webpack.HotModuleReplacementPlugin());
                let compiler = webpack(config);
                let assetsConfig = {
                    noInfo: true,
                    publicPath: config.output.publicPath,
                    stats: {
                        colors: true
                    }
                };
                if (process.platform === 'darwin') {
                    assetsConfig.watchOptions = {
                        aggregateTimeout: 1000,
                        poll: false,
                        useFsEvents: true,
                        watch: true
                    };
                } else if (process.platform !== 'win32') {
                    assetsConfig.watchOptions = {
                        aggregateTimeout: 300,
                        poll: true,
                        watch: true
                    };
                }
                assetsConfig = mergeWith(assetsConfig, this.config.packer.assets);
                let assets = mergeWith(assetsConfig, (this.config.packer && this.config.packer.devMiddleware) || {});
                let hot = mergeWith({
                    publicPath: config.output.publicPath
                }, (this.config.packer && this.config.packer.hotMiddleware) || {});

                this.hapiServer.register({
                    register: require('hapi-webpack-plugin'),
                    options: {
                        compiler,
                        assets,
                        hot
                    }
                }, err => {
                    if (err) {
                        reject(err);
                    } else {
                        this.hotReload = true;
                        resolve(true);
                    }
                });
            } else {
                // @TODO implement lasso hot reload
            }
        });
    };

    HttpServerPort.prototype.stop = function stop() {
        this.socketServer && this.socketServer.stop();
        return new Promise((resolve, reject) => {
            this.hapiServer.stop((err) => {
                return err ? reject(err) : Promise.resolve()
                    .then(() => parent && parent.prototype.stop.call(this))
                    .then(resolve)
                    .catch(reject);
            });
        });
    };

    HttpServerPort.prototype.status = function status() {
        return {
            port: this.hapiServer.info && this.hapiServer.info.port
        };
    };

    return HttpServerPort;
};
